### PR TITLE
Support protobuf separately to `GRPCPayload` for the client

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

To support payloads other than `SwiftProtobuf.Message` we required that
all messages conform to `GRPCPayload`. For protobuf messages we added
`GRPCProtobufPayload` which provides a default implemenation of
`GRPCPayload` for protobuf messages. We generated this conformance for
all protobuf messages we saw. This lead to a number issues and
workarounds including: #738, #778, #801, #837, #877, #881.

The intention is to continue to support `GRPCPayload` in addition to
protobuf, however, support for protobuf will not be via the
`GRPCProtobufPayload` protocol.

This PR builds on #886 by increasing the surface area of the client APIs
so that they are not constrained to `GRPCPayload`. The surface API now
has variants for `GRPCPayload` and `SwiftProtobuf.Message`. Internally
the client deals with serializers and deserializers.

Modifications:

- `GRPCClientChannelHandler` and `GRPCClientStateMachine` are no longer
  generic over a request and response type, rather they deal with the
  serialzed version of requests and response (i.e. `ByteBuffer`s) and
  defer the (de/)serialization to a separate handler.
- Added `GRCPClientCodecHandler` to handle (de/)serialization of
  messages
- Clients are no longer constrained to having their request/response
  payloads conform to `GRPCPayload`
- Conformance to `GRPCProtobufPayload` is no longer generated and the
  protocol is deprecated and has no requirements.
- Drop the 'GenerateConformance' option from the codegen since it is no
  longer required
- Reintroduce a filter to the codegen so that we only consider files
  which contain services, this avoids generating empty files
- Regenerate code where necessary

Result:

- `GRPCProtobufPayload` is no longer required

### CODE REVIEW

**Comment by glbrntt** [comment]
The non-generated changes are in: https://github.com/grpc/grpc-swift/commit/7a588894617e1d4630be330d00d79f444e487fe4

File: Sources/GRPC/ClientCalls/BidirectionalStreamingCall.swift [hunk 1]
```diff
@@ -147,16 +144,20 @@ public final class BidirectionalStreamingCall<
 }
 
 extension BidirectionalStreamingCall {
-  internal static func makeOnHTTP2Stream(
+  internal static func makeOnHTTP2Stream<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/BidirectionalStreamingCall.swift:147]: Nit: single letter type names are a bit sad, maybe consider `Serialize`/`Deserializer`?

File: Sources/GRPC/ClientCalls/BidirectionalStreamingCall.swift [hunk 2]
```diff
@@ -167,12 +168,14 @@ extension BidirectionalStreamingCall {
     return BidirectionalStreamingCall(transport: transport, options: callOptions)
   }
 
-  internal static func make(
+  internal static func make<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/BidirectionalStreamingCall.swift:171]: Same note here.

File: Sources/GRPC/ClientCalls/ClientCallTransport.swift [hunk 3]
```diff
@@ -148,14 +148,16 @@ internal class ChannelTransport<Request: GRPCPayload, Response: GRPCPayload> {
     channelProvider(self, channelPromise)
   }
 
-  internal convenience init(
+  internal convenience init<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/ClientCallTransport.swift:151]: Same note about type names here.

File: Sources/GRPC/ClientCalls/ClientStreamingCall.swift [hunk 4]
```diff
@@ -152,16 +149,20 @@ public final class ClientStreamingCall<
 }
 
 extension ClientStreamingCall {
-  internal static func makeOnHTTP2Stream(
+  internal static func makeOnHTTP2Stream<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/ClientStreamingCall.swift:152]: Same note about type names here.

File: Sources/GRPC/ClientCalls/ClientStreamingCall.swift [hunk 5]
```diff
@@ -171,11 +172,13 @@ extension ClientStreamingCall {
     return ClientStreamingCall(response: responsePromise.futureResult, transport: transport, options: callOptions)
   }
 
-  internal static func make(
+  internal static func make<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/ClientStreamingCall.swift:175]: Same note about type names here.

File: Sources/GRPC/ClientCalls/ServerStreamingCall.swift [hunk 6]
```diff
@@ -93,16 +90,20 @@ public final class ServerStreamingCall<
 }
 
 extension ServerStreamingCall {
-  internal static func makeOnHTTP2Stream(
+  internal static func makeOnHTTP2Stream<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/ServerStreamingCall.swift:93]: Same note about type names here.

File: Sources/GRPC/ClientCalls/ServerStreamingCall.swift [hunk 7]
```diff
@@ -113,12 +114,14 @@ extension ServerStreamingCall {
     return ServerStreamingCall(transport: transport, options: callOptions)
   }
 
-  internal static func make(
+  internal static func make<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/ServerStreamingCall.swift:117]: Same note about type names here.

File: Sources/GRPC/ClientCalls/UnaryCall.swift [hunk 8]
```diff
@@ -100,16 +97,20 @@ public final class UnaryCall<
 }
 
 extension UnaryCall {
-  internal static func makeOnHTTP2Stream(
+  internal static func makeOnHTTP2Stream<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/UnaryCall.swift:100]: Same note about type names here.

File: Sources/GRPC/ClientCalls/UnaryCall.swift [hunk 9]
```diff
@@ -119,11 +120,13 @@ extension UnaryCall {
     return UnaryCall(response: responsePromise.futureResult, transport: transport, options: callOptions)
   }
 
-  internal static func make(
+  internal static func make<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientCalls/UnaryCall.swift:123]: Same note about type names here.

File: Sources/GRPC/ClientConnection.swift [hunk 10]
```diff
@@ -143,19 +144,24 @@ public class ClientConnection {
   }
 }
 
-// Note: documentation is inherited.
+// MARK: - Unary
+
 extension ClientConnection: GRPCChannel {
-  public func makeUnaryCall<Request: GRPCPayload, Response: GRPCPayload>(
+  private func makeUnaryCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientConnection.swift:150]: Same note about type names here.

File: Sources/GRPC/ClientConnection.swift [hunk 11]
```diff
@@ -143,19 +144,24 @@ public class ClientConnection {
...
+    serializer: S,
+    deserialiser: D,
```
Comment (Lukasa) [inline Sources/GRPC/ClientConnection.swift:152]: Inconsistent spelling, use `z`.

Comment (Lukasa) [reply Sources/GRPC/ClientConnection.swift:152]: This is pervasive ("find on page" shows 32 instances) so I think you'll need to do a quick search for this in the diff and fix it everywhere.

File: Sources/GRPC/ClientConnection.swift [hunk 12]
```diff
@@ -186,18 +230,52 @@ extension ClientConnection: GRPCChannel {
...
+      path: path,
+      callOptions: callOptions
+    )
+  }
+}
+
+// MARK: - Server Streaming
+
+extension ClientConnection {
+  private func makeServerStreamingCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientConnection.swift:263]: Same note about type names here.

File: Sources/GRPC/ClientConnection.swift [hunk 13]
```diff
@@ -166,16 +172,54 @@ extension ClientConnection: GRPCChannel {
...
+      request: request,
+      callOptions: callOptions
+    )
+  }
+}
+
+// MARK: - Client Streaming
+
+extension ClientConnection {
+  private func makeClientStreamingCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientConnection.swift:209]: Same note about type names here.

File: Sources/GRPC/ClientConnection.swift [hunk 14]
```diff
@@ -209,17 +287,59 @@ extension ClientConnection: GRPCChannel {
...
+      callOptions: callOptions,
+      handler: handler
+    )
+  }
+}
+
+// MARK: - Bidirectional Streaming
+
+extension ClientConnection {
+  private func makeBidirectionalStreamingCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/ClientConnection.swift:328]: Same note about type names here.

File: Sources/GRPC/FakeChannel.swift [hunk 15]
```diff
@@ -171,3 +222,87 @@ extension FakeChannel {
     )
   }
 }
+
+extension FakeChannel {
+  private func makeUnaryCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/FakeChannel.swift:227]: Same note about type names here.

File: Sources/GRPC/FakeChannel.swift [hunk 16]
```diff
@@ -171,3 +222,87 @@ extension FakeChannel {
...
+      callOptions: callOptions,
+      logger: self.logger
+    )
+
+    call.send(self.makeRequestHead(path: path, callOptions: callOptions), request: request)
+
+    return call
+  }
+
+  private func makeClientStreamingCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/FakeChannel.swift:247]: Same note on type names here.

File: Sources/GRPC/FakeChannel.swift [hunk 17]
```diff
@@ -171,3 +222,87 @@ extension FakeChannel {
...
+      callOptions: callOptions,
+      logger: self.logger
+    )
+
+    call.sendHead(self.makeRequestHead(path: path, callOptions: callOptions))
+
+    return call
+  }
+
+  private func makeServerStreamingCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/FakeChannel.swift:266]: Same note on type names here.

File: Sources/GRPC/FakeChannel.swift [hunk 18]
```diff
@@ -171,3 +222,87 @@ extension FakeChannel {
...
+      logger: self.logger,
+      responseHandler: handler
+    )
+
+    call.send(self.makeRequestHead(path: path, callOptions: callOptions), request: request)
+
+    return call
+  }
+
+  private func makeBidirectionalStreamingCall<S: MessageSerializer, D: MessageDeserializer>(
```
Comment (Lukasa) [inline Sources/GRPC/FakeChannel.swift:288]: Same note on type names here.

File: Sources/GRPC/ClientConnection.swift (hunk 11 above)
Comment (glbrntt) [reply Sources/GRPC/ClientConnection.swift]: Agh, thanks, good spot.

**Pull request merged**